pkgs/build-support/rust/build-rust-package/default.nix

```diff
@@ -17,66 +17,56 @@
-, checkFeatures ? buildFeatures
-, useNextest ? false
-, auditable ? true
-
-, depsExtraArgs ? {}
-
-# Toggles whether a custom sysroot is created when the target is a .json file.
-, __internal_dontAddSysroot ? false
-
-# Needed to `pushd`/`popd` into a subdir of a tarball if this subdir
-# contains a Cargo.toml, but isn't part of a workspace (which is e.g. the
-# case for `rustfmt`/etc from the `rust-sources).
-# Otherwise, everything from the tarball would've been built/tested.
-, buildAndTestSubdir ? null
-, ... } @ args:
-
-assert cargoVendorDir == null && cargoLock == null
-    -> !(args ? cargoSha256 && args.cargoSha256 != null) && !(args ? cargoHash && args.cargoHash != null)
-    -> throw "cargoSha256, cargoHash, cargoVendorDir, or cargoLock must be set";
-assert buildType == "release" || buildType == "debug";
+let rustOverride =
+
+finalAttrs: previousAttrs:
 
 let
+  cargoPatches = finalAttrs.cargoPatches or [];
+  buildType = finalAttrs.buildType or "release";
+  checkType = finalAttrs.checkType or buildType;
+  cargoLock = finalAttrs.cargoLock or null;
+  cargoVendorDir = finalAttrs.cargoVendorDir or null;
+  buildNoDefaultFeatures = finalAttrs.buildNoDefaultFeatures or false;
+  checkNoDefaultFeatures = finalAttrs.checkNoDefaultFeatures or buildNoDefaultFeatures;
+  buildFeatures = finalAttrs.buildFeatures or [ ];
+  checkFeatures = finalAttrs.checkFeatures or buildFeatures;
+  useNextest = finalAttrs.useNextest or false;
+  auditable = finalAttrs.auditable or true;
+  depsExtraArgs = finalAttrs.depsExtraArgs or { };
+
+  # Toggles whether a custom sysroot is created when the target is a .json file.
+  __internal_dontAddSysroot = finalAttrs.__internal_dontAddSysroot or false;
+
+  # Needed to `pushd`/`popd` into a subdir of a tarball if this subdir
+  # contains a Cargo.toml, but isn't part of a workspace (which is e.g. the
+  # case for `rustfmt`/etc from the `rust-sources).
+  # Otherwise, everything from the tarball would've been built/tested.
+  buildAndTestSubdir = finalAttrs.buildAndTestSubdir or null;
 
   cargoDeps =
+    assert cargoVendorDir == null && cargoLock == null
+        -> !(finalAttrs.cargoSha256 or null != null) && !(finalAttrs.cargoHash or null != null)
+        -> throw "One of cargoSha256, cargoHash, cargoVendorDir or cargoLock must be set";
     if cargoVendorDir != null then null
     else if cargoLock != null then importCargoLock cargoLock
     else fetchCargoTarball ({
-      inherit src srcs sourceRoot preUnpack unpackPhase postUnpack cargoUpdateHook;
-      name = cargoDepsName;
+      src = finalAttrs.src or "";
+      srcs = finalAttrs.srcs or null;
+      sourceRoot = finalAttrs.sourceRoot or null;
+      preUnpack = finalAttrs.preUnpack or null;
+      unpackPhase = finalAttrs.unpackPhase or null;
+      # TODO using previousAttrs here as we otherwise trigger rebuilds for all
+      # FOD fetcher users as finalAttrs.postUnpack is prefixed below.
+      postUnpack = previousAttrs.postUnpack or null;
```

Comment:
Side effect of #221093. This is not ideal; I guess one solution would be to accept the rebuilds, but maybe there is a different approach here.

Reply:
Attributes with `null` values are completely removed from the `.drv`, so this doesn't look like a mass rebuild to me.

Reply:
Here is a concrete example: 

With this PR in its current state (48e9de7c830295a6cb3b765448995a56bda8dd40):

```console
 $ nix-instantiate -A cargo
/nix/store/1977acij72524vy0c8ij19ng7nzrf6yv-cargo-1.68.2.drv
```
After applying this diff
```diff
diff --git a/pkgs/build-support/rust/build-rust-package/default.nix b/pkgs/build-support/rust/build-rust-package/default.nix
index f8e1eef2e93..6ca8a730d59 100644
--- a/pkgs/build-support/rust/build-rust-package/default.nix
+++ b/pkgs/build-support/rust/build-rust-package/default.nix
@@ -58,7 +58,7 @@ let
       unpackPhase = finalAttrs.unpackPhase or null;
       # TODO using previousAttrs here as we otherwise trigger rebuilds for all
       # FOD fetcher users as finalAttrs.postUnpack is prefixed below.
-      postUnpack = previousAttrs.postUnpack or null;
+      postUnpack = finalAttrs.postUnpack or null;
       cargoUpdateHook = finalAttrs.cargoUpdateHook or "";
       # Name for the vendored dependencies tarball
       name = finalAttrs.cargoDepsName or finalAttrs.name or "${finalAttrs.pname}-${finalAttrs.version}";
```
we get
```console
 $ nix-instantiate -A cargo
/nix/store/3msgx9jda9kc6hvfl7vgf9b8nsbm5g6y-cargo-1.68.2.drv
```
Diff:
```console
 $ nix-diff /nix/store/1977acij72524vy0c8ij19ng7nzrf6yv-cargo-1.68.2.drv /nix/store/3msgx9jda9kc6hvfl7vgf9b8nsbm5g6y-cargo-1.68.2.drv
- /nix/store/1977acij72524vy0c8ij19ng7nzrf6yv-cargo-1.68.2.drv:{out}
+ /nix/store/3msgx9jda9kc6hvfl7vgf9b8nsbm5g6y-cargo-1.68.2.drv:{out}
• The input derivation named `cargo` differs
  - /nix/store/qcsj2rh9k0dgig555kb12vx6zrzsw2nr-cargo.drv:{out}
  + /nix/store/fmvz39xbarfgssly8j6lzjp5dcxpgwxb-cargo.drv:{out}
  • The input derivation named `cargo-auditable-0.6.1` differs
    - /nix/store/7gkiqcx8qigih6pzga18ps6w5dpv3n33-cargo-auditable-0.6.1.drv:{out}
    + /nix/store/lnxzsh3ij88gp0vzj0628p1dn4gd2irj-cargo-auditable-0.6.1.drv:{out}
    • The input derivation named `cargo-auditable-0.6.1-vendor.tar.gz` differs
      - /nix/store/n67n6kbyd0ch4v88wwra9fy57p6mr263-cargo-auditable-0.6.1-vendor.tar.gz.drv:{out}
      + /nix/store/wkvcmy569xq0lgwbfvismc7f1lwb8qhb-cargo-auditable-0.6.1-vendor.tar.gz.drv:{out}
      • The environments do not match:
          + postUnpack=eval "$cargoDepsHook"

export RUST_LOG=

    • Skipping environment comparison
  • Skipping environment comparison
• Skipping environment comparison
```
Does that demonstrate the problem more clearly?